### PR TITLE
LIA preprocessing: Rewrite div/mod constraints in the input formula

### PR DESCRIPTION
**Pull request by blishko**
This PR adds step for rewriting any div/mod constraints in the input formula to the theory preprocessing.
This step is added for `LIALogic` while for `LRALogic` it is basically a no-op.
After this change, OpenSMT will be able to decide satisfiability of formulas with div/mod constraints (if it can decide it for the rewritten pure LIA formula).

The rewriter is already present in the code (added in #221). The rewriting works as follows:
1. For any `div/mod` constraint `t div N` (or `t mod N`) we introduce two variables `d` and `m` representing values of the terms `t div N` and `t mod N`, respectively. 
2. Then a formula `F` is rewritten to `F'` such that `F' = F[t div N -> d; t mod N -> m] and t = d * N + m and 0 <= m <= |N| - 1`.
3. `F'`is then equisatisfiable with `F`. Any model of `F` can be extended to the model of `F'` by setting the value of `d` to the value of `t div N` (and the same for `m`). In the other direction, any model of `F'` is a model if `F` (ignoring the fresh variables `d` and `m`).

### CODE REVIEW

**Review by aehyvari — COMMENTED** [review]
This looks quite reasonable.  Should we include some regression-style unit tests?

**Comment by blishko** [comment]
Yeah, we probably should. I will add some basic ones. If you have any suggestions, let me know.

**Comment by blishko** [comment]
I had to update two assertions regarding arithmetic equalities elimination to use `isNumVarLike` instead of `isNumVarOrIte` to handle also `div/mod` terms.

I also added some very basic regression tests.

**Comment by blishko** [comment]
Given that we now have `RDLogic` and `IDLogic` that also use the `LATheory` template, I set the default for `rewriteDivMod` to do nothing and only implemented the specialization for `LIALogic`.
Before the template was undefined and specializations were defined separately for `LIALogic` and `LRALogic`, but this meant adding the "do nothing" definition for each new Logic separately. Having the default seems like a better option.

**Comment by aehyvari** [comment]
As discussed here's my suggestion on dealing with the div/mod question so that it won't bite us back with combinations of linear integer arithmetic and uninterpreted functions, where it would be possible for the user to define functions called `div` and `mod`.   There's also a fix for the nested `div`s and `mod`s.

**Comment by aehyvari** [comment]
So basically the idea is that while working over the API the functions mkIntDiv and mkIntMod are immediately available.  The feature needs to be specifically enabled only when the features are accessed through the automatic symbol resolution, such as when called from Interpret.  This is then implemented on `Interpret`: it checks the status from `SMTConfig` every time it sees an `(assert ...)` statement in the AST.  Do you think this is practical enough?

File: src/rewriters/Rewriter.h [hunk 1]
```diff
@@ -57,12 +58,10 @@ class Rewriter {
...
             }
             PTRef newTerm = needsChange ? logic.insertTerm(term.symb(), newArgs) : currentRef;
-            if (needsChange) {
-                substitutions.insert(currentRef, newTerm);
-            }
             // The reference "term" has now been possibly invalidated! Do not access it anymore!
+
             PTRef rewritten = cfg.rewrite(newTerm);
-            if (rewritten != newTerm) {
+            if (rewritten != newTerm or needsChange) {
```
Comment (blishko) [inline src/rewriters/Rewriter.h:64]: Nice! This is simpler and more elegant that the solutions we discussed before.

**Review by blishko — COMMENTED** [review]
This looks good. I'm fine with the option for allowing `div/mod` from the interpreter and the fix of the `Rewriter` is very nice.

**Comment by blishko** [comment]
@aehyvari, from my perspective, your changes are great. 
I cannot approve PR that I have started, so you will have to do that, if you are also satisfied with this.

**Comment by blishko** [comment]
I believe your changes in the `Substitutor` also resolve the main problem I raised in #271.

**Review by aehyvari — APPROVED** [review]

**Pull request merged**